hail/src/main/scala/is/hail/io/tabix/TabixReader.scala

```diff
@@ -0,0 +1,346 @@
+package is.hail.io.tabix
+
+import is.hail.HailContext
+import is.hail.io.compress.BGzipInputStream
+import is.hail.utils._
+
+import htsjdk.tribble.util.{ParsingUtils, TabixUtils}
+import org.apache.{hadoop => hd}
+import org.apache.hadoop.io.compress.SplitCompressionInputStream
+
+import java.io.InputStream
+import java.nio.{ByteBuffer, ByteOrder}
+import java.util.Arrays
+
+import scala.collection.mutable.HashMap
+import scala.language.implicitConversions
+
+// Helper data classes
+
+case class Tabix(
+  val format: Int,
+  val colSeq: Int,
+  val colBeg: Int,
+  val meta: Int,
+  val seqs: Array[String],
+  val chr2tid: HashMap[String, Int],
+  val indicies: Array[(HashMap[Int, Array[TbiPair]], Array[Long])]
+)
+
+case class TbiPair(var _1: Long, var _2: Long) extends java.lang.Comparable[TbiPair] {
+  @Override
+  def compareTo(other: TbiPair) = TbiOrd.compare(this, other)
+}
+
+object TbiPair {
+  implicit def tup2Pair(t: (Long, Long)): TbiPair = new TbiPair(t._1, t._2)
+}
+
+object TbiOrd extends Ordering[TbiPair] {
+  def compare(u: TbiPair, v: TbiPair) = if (u._1 == v._1) {
+    0
+  } else if ((u._1 < v._1) ^ (u._1 < 0) ^ (v._1 < 0)) {
+    -1
+  } else {
+    1
+  }
+
+  def less64(u: Long, v: Long) = (u < v) ^ (u < 0) ^ (v < 0)
+}
+
+// Tabix reader
+// Tabix file format is described here: https://samtools.github.io/hts-specs/tabix.pdf
+
+object TabixReader {
+  val MaxBin: Int = 37450
+  val TadLidxShift: Int = 14
+  val DefaultBufferSize: Int = 1000
+  val Magic: Array[Byte] = Array(84, 66, 73, 1) // "TBI\1"
+
+  def readInt(is: InputStream): Int = {
+    val buf = new Array[Byte](4)
```

Comment:
The performance of this code is going to be pretty poor. 

Let's do this without allocation, like AbstractBinaryReader.readInt